src/types/stack.rs

```diff
@@ -91,6 +91,75 @@ impl StackFrame {
             red::OpcodeHandlers::Run(opcode, self.0.context, &mut self.0, ptr, ptr::null_mut());
         }
     }
+
+    /// Captures stack frame state.
+    ///
+    /// Use its returned value
+    /// with [rewind](Self::rewind) to restore stack.
+    pub fn state(&self) -> StackState {
+        StackState {
+            code: self.0.code,
+            data: self.0.data,
+            data_type: self.0.dataType,
+        }
+    }
+
+    /// Allows to rewind the stack after having read function arguments,
+    /// when detoured function need to be called.
+    ///
+    /// # Safety
+    /// The state must be saved **before** reading arguments.
+    ///
+    /// The state must be restored **before** calling callback,
+    /// once arguments have been read.
+    ///
+    /// Stack arguments should **neither** be _partially_ read,
+    /// **nor** _partially_ restored.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use red4ext_rs::{hooks, SdkEnv, types::{CName, EntityId, StackFrame, IScriptable}, VoidPtr};
+    /// # hooks! {
+    /// #    static ADD_HOOK: fn(i: *mut IScriptable, f: *mut StackFrame, a3: VoidPtr, a4: VoidPtr) -> ();
+    /// # }
+    /// # fn attach_my_hook(env: &SdkEnv, addr: unsafe extern "C" fn(i: *mut IScriptable, f: *mut StackFrame, a3: VoidPtr, a4: VoidPtr)) {
+    /// #     unsafe { env.attach_hook(ADD_HOOK, addr, detour) };
+    /// # }
+    /// # fn should_detour(event_name: CName) -> bool { false }
+    ///
+    /// unsafe extern "C" fn detour(
+    ///     i: *mut IScriptable,
+    ///     f: *mut StackFrame,
+    ///     a3: VoidPtr,
+    ///     a4: VoidPtr,
+    ///     cb: unsafe extern "C" fn(i: *mut IScriptable, f: *mut StackFrame, a3: VoidPtr, a4: VoidPtr),
+    /// ) {
+    ///     let frame = &mut *f;
+    ///
+    ///     // stack must be saved before reading stack function parameters
+    ///     let state = frame.state();
+    ///     
+    ///     // assuming our function accepts these 3 parameters
+    ///     let event_name: CName = StackFrame::get_arg(frame);
+    ///     let entity_id: EntityId = StackFrame::get_arg(frame);
+    ///     let emitter_name: CName = StackFrame::get_arg(frame);
+    ///
+    ///     if should_detour(event_name) {
+    ///         // do something else...
+    ///     } else {
+    ///         // since we've read stack function arguments,
+    ///         // stack must be rewinded before callback.
+    ///         frame.rewind(state);
+    ///         cb(i, f, a3, a4);
+    ///     }
+    /// }
+    /// ```
+    pub unsafe fn rewind(&mut self, state: StackState) {
```

Comment:
the name `rewind` suggests that the stack itself is being popped and this is not doing anything like that, it just resets the state of function arguments, so I think it should not be called `rewind`
```suggestion
    pub unsafe fn reset_args(&mut self, state: StackState) {
```